links/src/commonMain/kotlin/utils/decoders.kt

```diff
@@ -18,61 +18,192 @@
 package com.splendo.kaluga.links.utils
 
 import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.KSerializer
 import kotlinx.serialization.SerializationException
 import kotlinx.serialization.descriptors.SerialDescriptor
-import kotlinx.serialization.encoding.AbstractDecoder
 import kotlinx.serialization.encoding.CompositeDecoder
+import kotlinx.serialization.encoding.Decoder
 import kotlinx.serialization.modules.EmptySerializersModule
 import kotlinx.serialization.modules.SerializersModule
 
-internal fun <T> decodeFromList(list: List<Any>, deserializer: DeserializationStrategy<T>): T {
-    val decoder = LinksDecoder(ArrayDeque(list))
-    return decoder.decodeSerializableValue(deserializer)
-}
+internal fun <T> decodeFromMap(fields: Map<String, List<String>>, serializer: KSerializer<T>): T = MapDecoder(fields).decodeSerializableValue(serializer)
 
-internal class LinksDecoder(private val list: ArrayDeque<Any>, var elementsCount: Int = 0) : AbstractDecoder() {
-    private var elementIndex = 0
+/**
```

Comment:
a rather big changeset but now parsing supports inline classes, default values and random field order. also lists don't have to specify size in url any more